quinn/src/lib.rs

```diff
@@ -546,6 +323,266 @@ impl EndpointInner {
+                    if let Some(x) = pending.incoming_streams_reader.take() {
+                        x.notify();
+                    }
+                }
+                StreamReadable { stream } => {
+                    let pending = self.pending.get_mut(&ch).unwrap();
+                    if let Some(reader) = pending.blocked_readers.remove(&stream) {
+                        reader.notify();
+                    }
+                }
+                StreamAvailable { directionality } => {
+                    let pending = self.pending.get_mut(&ch).unwrap();
+                    let queue = match directionality {
+                        Directionality::Uni => &mut pending.uni_opening,
+                        Directionality::Bi => &mut pending.bi_opening,
+                    };
+                    while let Some(connection) = queue.pop_front() {
+                        if let Some(id) = self.inner.open(ch, directionality) {
+                            let _ = connection.send(Ok(id));
+                        } else {
+                            queue.push_front(connection);
+                            break;
+                        }
+                    }
+                }
+                StreamFinished { stream } => {
+                    let _ = self
+                        .pending
+                        .get_mut(&ch)
+                        .unwrap()
+                        .finishing
+                        .remove(&stream)
+                        .unwrap()
+                        .send(None);
+                }
+            }
+        }
+    }
+
+    fn poll_incoming(&mut self, wrapped: Rc<RefCell<EndpointInner>>) {
+        while let Ok(Async::Ready(())) = self.incoming.poll_ready() {
+            if let Some(ch) = self.buffered_incoming.pop_front() {
+                self.incoming
+                    .start_send(NewConnection::new(wrapped.clone(), ch))
+                    .unwrap();
+                self.inner.accept();
+            } else {
+                break;
+            }
+        }
+    }
+
+    fn poll_send(&mut self) -> Result<bool, io::Error> {
+        let mut blocked = false;
+        if let Some(ref x) = self.outgoing {
+            match self.socket.poll_send(&x.destination, x.ecn, &x.packet) {
+                Ok(Async::Ready(_)) => {
+                    self.outgoing = None;
+                }
+                Ok(Async::NotReady) => {
+                    blocked = true;
+                }
+                Err(ref e) if e.kind() == io::ErrorKind::PermissionDenied => {
+                    blocked = true;
+                }
+                Err(e) => {
+                    return Err(e);
+                }
+            }
+        }
+        Ok(blocked)
+    }
+
+    fn poll_transmit(&mut self, now: Instant) -> Result<bool, io::Error> {
+        let mut sent = 0;
+        let mut keep_going = false;
+        while let Some(x) = self.inner.poll_transmit(now) {
```

Comment:
`while let Some(x) = self.outgoing.take().or_else(|| self.inner.poll_transmit(now))` would render `poll_send` unnecessary; pre-existing issue, though, so not a blocker.